File: spot_driver/scripts/servo.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import Float32MultiArray
import Adafruit_PCA9685

class Servo:

    def __init__(self):
        # Subscribe 12 joint angles [rad]
        self.sub = rospy.Subscriber("joint", Float32MultiArray, self.jointCallback, queue_size=10)

        # Initialise the PCA9685 using desired address & bus
        self.driver = Adafruit_PCA9685.PCA9685(address = 0x40, busnum = 1)
        self.driver.set_pwm_freq(60)

        # Load joint config from parameter server
        self.limit = rospy.get_param('limit')

    def jointCallback(self, msg):
        print(msg.data[0])
        #self.driver.set_pwm(0, 0, int(self.limit[0]['zero']) + int(450 * msg.data[0] / 3.14159))
        #self.driver.set_pwm(1, 0, int(self.limit[1]['zero']) + int(450 * msg.data[1] / 3.14159))
        #self.driver.set_pwm(2, 0, int(self.limit[2]['zero']) + int(450 * msg.data[2] / 3.14159))
        #self.driver.set_pwm(3, 0, int(self.limit[3]['zero']) + int(450 * msg.data[3] / 3.14159))
        for i in range(12):
            self.driver.set_pwm(i, 0, int(self.limit[i]['zero']) + int(450 * msg.data[i] / 3.14159))
            #self.driver.set_pwm(i, 0, int(self.limit[i]['zero']) + 450) )

if __name__ == '__main__':
    rospy.init_node('servo')
    Servo()
    rospy.spin()

File: spot_driver/src/servo.cpp
#include "servo.h"

Servo::Servo() {
  // Subscribe joint angles
  subJoint = nh.subscribe<std_msgs::Float32MultiArray>("angle", 1, &Servo::jointCallback, this);

  // Set timer
  timer = nh.createTimer(ros::Duration(0.1), &Servo::timerCallback, this);

  // Setup pca9685 (bus, address)
  driver = new PCA9685(1, 0x40);
  driver->setPWMFreq(60);

  // Get joint angle limitations (upper & lower boundary)
  nh.getParam("joint", joint);

  // Initial servo position
  for(int i = 0; i < 12; i++) {
    theta[i] = 0;
  }
}

Servo::~Servo() {
  delete driver;
}

void Servo::jointCallback(const std_msgs::Float32MultiArray::ConstPtr &joint) {
  for(int i = 0; i < 12; i++) {
    theta[i] = joint->data[i];
  }
}

void Servo::timerCallback(const ros::TimerEvent& event){
  for(int i = 0; i < 12; i++) {
    // channel = joint index + 1
    driver->setPWM(static_cast<int>( this->joint[i]["id"] ) + 1, 0,
                   static_cast<int>( this->joint[i]["center"] ) +
                   static_cast<int>( this->joint[i]["direction"] ) * int(450 * theta[i] / 3.14159) );
  }
}

int main(int argc, char **argv) {
  ros::init(argc, argv, "servo");
  Servo servo;
  ros::spin();
}
